### PR TITLE
ENH: override halflogistic fit for free parameters

### PR DESCRIPTION
**Pull request by dschmitz89**
#### Reference issue
Related to #11782

#### What does this implement/fix?
Overrides the `fit` method for the half logistic distribution acc. to [this paper](https://www.researchgate.net/publication/253328579_Comparisons_of_Methods_of_Estimation_for_the_Half-Logistic_Distribution/citation/download). The case that `floc` is given by the user is also treated in the publication and left for a follow up intentionally to make review of this PR easier.

#### Additional information
In a nutshell: for data $(y_1, ..., y_n)$, we get

$$floc=y_{min}=min(y_1, ..., y_n)$$

For the scale $\sigma$:

$$ \sigma=f(\sigma)=y_{mean} - y_{min} - \frac{2}{n}\sum_{i=1}^n\frac{(y_i-y_{min})\exp(-(y_i-y_{min})/\sigma)}{1+\exp(-(y_i-y_{min})/\sigma)} $$

The paper proposed to solve this equation via the simple iterative sequence $\sigma_{i+1}=f(\sigma_i)$. In my tests, this worked quite well and better than rootfinding.

### CODE REVIEW

File: scipy/stats/_continuous_distns.py [hunk 1]
```diff
@@ -4242,6 +4247,43 @@ def _munp(self, n):
     def _entropy(self):
         return 2-np.log(2)
 
+    def fit(self, data, *args, **kwds):
+        if (isinstance(data, CensoredData)):
```
Comment (mdhaber) [inline scipy/stats/_continuous_distns.py:4251]: This isn't what you want. Please consider the `_call_super_mom` decorator instead. I think that will do the right thing here.

Comment (mdhaber) [reply scipy/stats/_continuous_distns.py:4251]: The generic test for censored data for all distributions only checks that `CensoredData` with very tight interval censoring and `CensoredData` with no actual censoring work approximately the same as regular arrays. There is not a check that `CensoredData` with more general censoring does something *different*. We don't need to add that now. With the new infrastructure I'll be working on shortly, this will be less of a concern because distributions won't be overriding the public `fit` method.

File: scipy/stats/tests/test_distributions.py [hunk 2]
```diff
@@ -1175,6 +1175,17 @@ def test_sf(self, x, ref):
...
+    @pytest.mark.parametrize("rvs_loc", [1e-5, 1e10])
+    @pytest.mark.parametrize("rvs_scale", [1e-2, 100, 1e8])
+    def test_fit_MLE_comp_optimizer(self, rvs_loc, rvs_scale):
+
+        rng = np.random.default_rng(6762668991392531563)
+        data = stats.halflogistic.rvs(loc=rvs_loc, scale=rvs_scale, size=1000,
+                                      random_state=rng)
+
+        _assert_less_or_close_loglike(stats.halflogistic, data,
+                                      stats.halflogistic.nnlf)
```
Comment (mdhaber) [inline scipy/stats/tests/test_distributions.py:1187]: For simplicity, since it's the default.

```suggestion
        _assert_less_or_close_loglike(stats.halflogistic, data)
```

**Review by mdhaber — COMMENTED** [review]
This looks correct, but there are a few suggestions.

Also, please run experiments like:
https://github.com/scipy/scipy/pull/16839#issuecomment-1217437632
and post the code/results.

I think the range of parameters given in that code is appropriate. I think you just need to replace `stats.lognorm` with `stats.halflogistic` (or define `dist = stats.halflogistic` at the top and use it throughout for ease of future modifications) and remove mentions of `s`.

If the experiments show that it's robust for a wide range of parameters and sample sizes, I think we can merge this pretty quickly.

File: scipy/stats/_continuous_distns.py [hunk 3]
```diff
@@ -4242,6 +4247,43 @@ def _munp(self, n):
...
+            floc = np.min(data)
+
+            # scale is solution to a fix point problem
+            # first, precompute shifted data and constants
+            mean_val = np.mean(data)
+            shifted_data = data - floc
+            mean_minus_min = mean_val - floc
+            n_samples = data.shape[0]
+
+            # heuristically found promising starting point
```
[the remark this answers is not in the document]
Comment (mdhaber) [reply scipy/stats/_continuous_distns.py:4272]: How about using the approximate value given in the paper?

image

**Comment by dschmitz89** [comment]
> This looks correct, but there are a few suggestions.
> 
> Also, please run experiments like: [#16839 (comment)](https://github.com/scipy/scipy/pull/16839#issuecomment-1217437632) and post the code/results.
> 
> I think the range of parameters given in that code is appropriate. I think you just need to replace `stats.lognorm` with `stats.halflogistic` (or define `dist = stats.halflogistic` at the top and use it throughout for ease of future modifications) and remove mentions of `s`.
> 
> If the experiments show that it's robust for a wide range of parameters and sample sizes, I think we can merge this pretty quickly.

I think the results are very convincing. This is still using my simple starting guess, not the approximation given in the paper. Will iron that out over the next days-

Time
Rel_Improvements
Rel_Regressions

<details>

```python
import numpy as np
from scipy import stats
from time import perf_counter
import matplotlib.pyplot as plt

rng = np.random.default_rng(2433006235492611347)
qrng = stats.qmc.Halton(5, seed=rng)

N = 1000
times_pr = np.full(N, np.nan)
times_super = np.full(N, np.nan)
nllfs_super = np.full(N, np.nan)
nllfs_pr = np.full(N, np.nan)

for i in range(N):
    print(i)
    sample = qrng.random()
    sample = stats.qmc.scale(sample, [-3, -3, -3, 1, -1], [2, 3, 3, 5, 1])[0]
    _, loc, scale, size = 10**sample[:4]
    size = int(size)

    dist = stats.halflogistic(loc=loc, scale=scale)
    rvs = dist.rvs(size=size, random_state=rng)

    try:
        tic = perf_counter()
        loc_fit, scale_fit = stats.halflogistic.fit(rvs)
        toc = perf_counter()
        nllf_pr = stats.halflogistic.nnlf((loc_fit, scale_fit), rvs)
        nllfs_pr[i] = nllf_pr
        times_pr[i] = toc-tic

    except Exception as e:
        print(f"Failure: {e}")

    try:
        tic = perf_counter()
        loc_fit, scale_fit = stats.halflogistic.fit(rvs, superfit=True)
        toc = perf_counter()
        nllf_super = stats.halflogistic.nnlf((loc_fit, scale_fit), rvs)
        nllfs_super[i] = nllf_super
        times_super[i] = toc-tic

    except Exception as e:
        print(f"Failure: {e}")

nllf_diff = (nllfs_pr - nllfs_super)/np.abs(nllfs_super)
nllf_diff = nllf_diff[np.isfinite(nllf_diff)]
nllf_diff_good = -nllf_diff[nllf_diff < 0]
nllf_diff_bad = nllf_diff[nllf_diff > 0]

bins = np.arange(-3, 1.5, 0.25)
plt.hist(np.log10(times_pr), bins=bins, alpha=0.5)
plt.hist(np.log10(times_super), bins=bins, alpha=0.5)
plt.legend(('PR', 'main'))
plt.xlabel('log10 of `fit` execution time (s)')
plt.ylabel('Frequency')
plt.title('Histogram of log10 of `fit` execution time (s)')
plt.savefig("./Time.png", dpi=300)

plt.clf()
plt.hist(np.log10(nllf_diff_good))
plt.xlabel('log10 of NLLF relative improvements')
plt.ylabel('Frequency')
plt.title('Histogram of log10 of NLLF relative improvements')
plt.savefig("Rel_Improvements.png", dpi=300)

plt.clf()
plt.hist(np.log10(nllf_diff_bad))
plt.xlabel('log10 of NLLF relative regression')
plt.ylabel('Frequency')
plt.title('Histogram of log10 of NLLF relative regression')
plt.savefig("Rel_Regressions.png", dpi=300)

```

</details>

File: scipy/stats/_continuous_distns.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (dschmitz89) [reply scipy/stats/_continuous_distns.py:4271]: This looks indeed useful, will try it out.

**Comment by mdhaber** [comment]
OK, so all improvements / no regressions. The original plot you posted looked like there was less mass in the histogram for the PR times - do you figure out why that was and fix it?

**Comment by dschmitz89** [comment]
> OK, so all improvements / no regressions. The original plot you posted looked like there was less mass in the histogram for the PR times - do you figure out why that was and fix it?
> 
> And would you like this to merge if it looks good, or would you like to add in the approximate MLE as a guess?

I had used the original x-axis range starting from 0.001, then I lowered it to `1e-6`. I will look into the approximate MLE this week still, no need to rush merging. I also have to still fix the Censored data.

**Comment by dschmitz89** [comment]
With the approximate MLE as initial guess for the scale, timings improve (compare to above). Likelihood improvements stay the same.

Time
Rel_Improvements

**Comment by dschmitz89** [comment]
Looks like some test failures are related:

```

FAILED stats/tests/test_fit.py::test_cont_fit[MLE-halflogistic-arg42] - AssertionError: 
Not equal to tolerance rtol=1e-10, atol=0
Different results fitting uncensored data wrapped as CensoredData: halflogistic: est=(0.0030875275950887726, 1.042744883380042) est1=(0.0030875107564566806, 1.0427304240258568)
Mismatched elements: 2 / 2 (100%)
Max absolute difference: 1.44593542e-05
Max relative difference: 1.38666268e-05
 x: array([0.003088, 1.04273 ])
 y: array([0.003088, 1.042745])
= 1 failed, 54682 passed, 2480 skipped, 256 xfailed, 9 xpassed, 3 warnings in 812.65s (0:13:32) =

```

File: scipy/stats/_continuous_distns.py [hunk 4]
```diff
@@ -4242,6 +4247,53 @@ def _munp(self, n):
     def _entropy(self):
         return 2-np.log(2)
 
+    @_call_super_mom
+    @inherit_docstring_from(rv_continuous)
+    def fit(self, data, *args, **kwds):
+        if (isinstance(data, CensoredData)) or kwds.pop('superfit', False):
```
Comment (mdhaber) [inline scipy/stats/_continuous_distns.py:4253]:
```suggestion
    def fit(self, data, *args, **kwds):
        if kwds.pop('superfit', False):
```

Please double-check, but I don't think either of these are needed. 

1. Because `fit` here has no docstring (and because `_call_super_mom` uses `wraps` to avoid getting in the way), this `fit` gets the docstring from `rv_continuous` without the `inherit_docstring_from` decorator.  (TBH it doesn't look like `inherit_docstring_from` does anything unless the docstring has `%(super)s` in the text.) In any case, you can verify that `from scipy.stats import halflogistic; help(halflogistic.fit)` does the right thing with and without the `inherit_docstring_from` decorator.

2. The input to `fit` can't be a `CensoredData` because if it is, the `_call_super_mom` wrapper takes care of it.

File: scipy/stats/_continuous_distns.py [hunk 5]
```diff
@@ -4242,6 +4247,53 @@ def _munp(self, n):
...
+        B = 2 * np.sum(alpha[1:] * sorted_data[1:])
+        C = 2 * np.sum(beta[1:] * sorted_data[1:]**2)
+        # starting guess
+        scale = ((B + np.sqrt(B**2 + 8 * n_observations * C))
+                 /(4 * n_observations))
+
+        # relative tolerance of fix point iterator
+        rtol = 1e-8
+        relative_residual = 1
+        shifted_mean = sorted_data.mean() #y_mean - y_min
```
Comment (mdhaber) [inline scipy/stats/_continuous_distns.py:4284]:
```suggestion
        shifted_mean = sorted_data.mean()  # y_mean - y_min
```
https://peps.python.org/pep-0008/#inline-comments

File: scipy/stats/_continuous_distns.py (hunk 4 above)
Comment (dschmitz89) [reply scipy/stats/_continuous_distns.py:4253]: CircleCI was complaning in the last run before about this, that's why I added it. But I am getting more and more confused with the intricacies here. The CI failure from https://github.com/scipy/scipy/pull/18695#issuecomment-1599409680 also seems to be gone now. Feel free to push whatever is needed regarding the docstrings and censored data to this branch to get this over the finish line if you can find time. Once more thanks for your advice and thorough reviews.

Comment (mdhaber) [reply scipy/stats/_continuous_distns.py:4253]: https://github.com/scipy/scipy/pull/18695#issuecomment-1599409680 was because https://github.com/scipy/scipy/pull/18695#discussion_r1233413979 hadn't been addressed. I suggested `inherit_docstring_from` and `extend_notes_in_docstring` to be considered depending on what you wanted to do with the docstring, not to just apply.

> Feel free to push whatever is needed regarding the docstrings and censored data to this branch to get this over the finish line if you can find time.

That would be as easy as pushing "Commit suggestion", but I would prefer not to here. (Every time I do, it slightly devalues the time already invested into review.) Do you plan on doing any more `fit` PRs? If so, I'd ask that you look into what's going on with these. Feel free to ask questions if you don't understand something.

File: scipy/stats/_continuous_distns.py [hunk 6]
```diff
@@ -4242,6 +4247,51 @@ def _munp(self, n):
...
+            return super().fit(data, *args, **kwds)
+
+        # location is the minimum of the sample ([1] Equation 2.3)
+        loc = np.min(data)
+
+        # scale is solution to a fix point problem ([1] 2.6)
+        # use approximate MLE as starting point ([1] 3.1)
+        n_observations = data.shape[0]
+        sorted_data = np.sort(data, axis=0)
+        p = np.arange(0, n_observations, 1)/(n_observations + 1)
```
Comment (mdhaber) [inline scipy/stats/_continuous_distns.py:4269]: Should this be
```suggestion
        p = np.arange(1, n_observations+1)/(n_observations + 1)
```
?

**Comment by dschmitz89** [comment]
The correction of the formula for `p` gave another performance improvement:

Time

**Review by mdhaber — APPROVED** [review]

**Comment by mdhaber** [comment]
The failures are unrelated. This looks like a robust improvement to `halflogistic.fit`.
